csp/middleware.py

```diff
@@ -50,35 +48,48 @@ def process_response(self, request, response):
             return response
 
         # Check for ignored path prefix.
-        prefixes = getattr(settings, 'CSP_EXCLUDE_URL_PREFIXES', ())
+        # TODO: Legacy setting
+        prefixes = getattr(
+            settings,
+            'CSP_EXCLUDE_URL_PREFIXES',
+            defaults.EXCLUDE_URL_PREFIXES,
+        )
         if request.path_info.startswith(prefixes):
             return response
 
         # Check for debug view
-        status_code = response.status_code
-        exempted_debug_codes = (
-            http_client.INTERNAL_SERVER_ERROR,
-            http_client.NOT_FOUND,
-        )
-        if status_code in exempted_debug_codes and settings.DEBUG:
+        if response.status_code in EXEMPTED_DEBUG_CODES and settings.DEBUG:
             return response
 
-        header = 'Content-Security-Policy'
-        if getattr(settings, 'CSP_REPORT_ONLY', False):
-            header += '-Report-Only'
-
-        if header in response:
+        existing_headers = {
+            header for header in HTTP_HEADERS if header in response
+        }
+        if len(existing_headers) == len(HTTP_HEADERS):
             # Don't overwrite existing headers.
             return response
 
-        response[header] = self.build_policy(request, response)
+        # These won't be ordered properly on Cpython 3.5 or below
+        # (out of support October 2020)
+        headers = defaultdict(list)
 
+        for csp, report_only in self.build_policy(request, response):
+            header = HTTP_HEADERS[int(report_only)]
+            if header in existing_headers:  # don't overwrite
+                continue
+            headers[header].append(csp)
+
+        for header, policies in headers.items():
+            # TODO: Design Decision do we want the optional whitespace?
+            # Maybe only in DEBUG mode?
+            response[header] = ', '.join(policies)
```

Comment:
Design Decision:  technically multiple of these headers are allowed I believe, but I'm not sure Django supports that.

See https://github.com/w3c/webappsec-csp/issues/215#:~:text=A%20server%20MUST%20NOT%20send,resource%20or%20with%20different%20resources.

If Django doesn't support it, we might want to validate that there's only one policy for each header.   If Django does support it and we want to support it, we may want to reconsider the "not overwriting existing headers" behaviour.

Reply:
I'm silly.  We support multiple already my merging them together.

Reply:
Doe we want to validate that they don't conflict?  Spec says a user agent is supposed to handle this, but also says we're supposed to avoid it.